pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -1449,6 +1449,15 @@ private ByteBuf processMessageChunk(ByteBuf compressedPayload, MessageMetadata m
         // discard message if chunk is out-of-order
         if (chunkedMsgCtx == null || chunkedMsgCtx.chunkedMsgBuffer == null
                 || msgMetadata.getChunkId() != (chunkedMsgCtx.lastChunkedMessageId + 1)) {
+            // Filter duplicated chunks instead of discard it.
+            if (chunkedMsgCtx == null || msgMetadata.getChunkId() <= chunkedMsgCtx.lastChunkedMessageId) {
```

Comment:
It seems like the above `if` is already handling the duplicated chunks, isn't it?

`|| msgMetadata.getChunkId() != (chunkedMsgCtx.lastChunkedMessageId + 1)`

Reply:
No. 
`msgMetadata.getChunkId() != (chunkedMsgCtx.lastChunkedMessageId + 1)` may mean chunk ID <= lastChunkedMessageId or chunk ID > lastChunkedMessageId + 1.
We only ignore the chunks we have received and call rewind if the chunk is out of order.

Reply:
I see.

So you are handling the first part here, `chunk ID <= lastChunkedMessageId`, without stopping the current msg chunking.

 I see that you also emitted this chunk expire logic here. can you explain why?

```
 if (expireTimeOfIncompleteChunkedMessageMillis > 0
                    && System.currentTimeMillis() > (msgMetadata.getPublishTime()
                            + expireTimeOfIncompleteChunkedMessageMillis)) {
                doAcknowledge(msgId, AckType.Individual, Collections.emptyMap(), null);
            } else {
                trackMessage(msgId);
            }
```

Reply:
What do you mean? Why should we change the expiration logic here?

Reply:
We have ignored the chunks whose msgMetadata.getChunkId() <= chunkedMsgCtx.lastChunkedMessageId, and it will not caught in a loop. Therefore, we should skip this expire logic here.

Reply:
I have reverted the previous modification. Feel free to ping me if there are other questions.

Reply:
Shouldn't we acknowledge the duplicated msgs(`msgMetadata.getChunkId() <= chunkedMsgCtx.lastChunkedMessageId`) to `ignore`?

Reply:
No, It has the same message ID. it will not ack the chunk message instead of the the single chunk.